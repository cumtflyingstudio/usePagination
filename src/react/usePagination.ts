import { useState, useReducer, useRef, useEffect } from "react";
interface IDataItem<T> {
  id: number | string;
  item: T;
}
type IDataList<T> = IDataItem<T>[];

interface IAction<T> {
  type: string;
  payload: IDataList<T>;
}
function reducer<T>(state: IDataList<T>, action: IAction<T>): IDataList<T> {
  let res = [] as any[];
  switch (action.type) {
    case "ADD":
      res = [...state, ...action.payload];
      break;
    case "RESET":
      res = [...action.payload];
      break;
    case "CLEAR":
      return [];
    default:
      return state;
  }
  // remove duplication
  let idSet = new Set();
  return res.filter((i) => {
    if (idSet.has(i.id)) {
      return false;
    } else {
      idSet.add(i.id);
      return true;
    }
  });
}
function formatItem<T>(item: T, idPropertyName) {
  const id = item?.[idPropertyName] ?? Math.random();
  return {
    id,
    item: item,
  };
}
/**
 *
 * @param paginationRequest 请求函数，返回需要请求到的数组
 * @example 
 * const  {
    data,
    loading,
    error,
    run,
    refresh,
    fetchData,
  } = usePagination((currPage) => request(currPage))
 */
const usePagination = <T>(
  paginationRequest: (currPage: number) => Promise<T[]>,
  option = {
    idPropertyName: "id",
    initialPage: 1,
    beforeAllRequest: (currPage: number) => {},
  }
) => {
  const [list, dispatch] = useReducer(reducer, [] as IDataList<T>);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);
  const {
    idPropertyName = "id",
    initialPage = 1,
    beforeAllRequest = () => {},
  } = option;
  const page = useRef(initialPage);
  const fetchData = async (type: string) => {
    setLoading(true);
    try {
      const dataList = await paginationRequest(page.current++);
      const payload = dataList.map((item) => formatItem(item, idPropertyName));
      dispatch({ type, payload });
      return payload;
    } catch (error) {
      setError(true);
    } finally {
      setLoading(false);
    }
  };
  const refresh = () => {
    page.current = initialPage;
    fetchData("RESET");
  };
  const run = (currPage: number = page.current) => {
    page.current = currPage;
    fetchData("ADD");
  };
  useEffect(() => {
    refresh();
  }, []);
  return {
    data: list.map((i) => i["item"]) as T[],
    loading,
    error,
    run,
    refresh,
    fetchData,
  };
};
export default usePagination;
export { usePagination };
