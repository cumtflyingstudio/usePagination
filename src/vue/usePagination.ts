function usePagination() {
  console.log('usePagination Vue will be released soon.');
}
export default usePagination;
