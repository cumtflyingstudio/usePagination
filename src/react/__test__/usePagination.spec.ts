import { describe, it, expect } from "vitest";
import { renderHook, act } from "@testing-library/react-hooks";
import { usePagination } from "../usePagination";
const mockPaginationRequest = (
  currPage: number,
  // odd -> 1000X, even -> 2000X
  id: number = (Number(!(currPage % 2)) + 1) * 10000 + currPage
) => {
  function createItem(currPage) {
    return {
      id: id++,
      name: `Page:${currPage}`,
    };
  }
  return Promise.resolve([createItem(currPage), createItem(currPage)]);
};
const setUpHook = <T>(
  paginationRequest: (currPage: number) => Promise<T[]>
) => {
  const renderHookRes = renderHook(() => usePagination(paginationRequest));
  return renderHookRes;
};
describe("usePagination", () => {
  it("should be defined", () => {
    const { result, rerender } = setUpHook(() => Promise.resolve([]));
    expect(result.current).toBeDefined();
  });
  it("init should be with refresh once", async () => {
    const { result, rerender } = setUpHook(mockPaginationRequest);
    expect(result.current.data).toEqual([]);
    await act(async () => {});
    expect(result.current.data).toEqual([
      { id: 10001, name: "Page:1" },
      { id: 10002, name: "Page:1" },
    ]);
  });
  it("test run once", async () => {
    const { result, rerender } = setUpHook(mockPaginationRequest);
    expect(result.current.data).toEqual([]);
    await act(async () => {
      await result.current.run();
    });
    expect(result.current.data).toEqual([
      { id: 10001, name: "Page:1" },
      { id: 10002, name: "Page:1" },
      { id: 20002, name: "Page:2" },
      { id: 20003, name: "Page:2" },
    ]);
  });
  it("test run twice", async () => {
    const { result, rerender } = setUpHook(mockPaginationRequest);
    expect(result.current.data).toEqual([]);
    await act(async () => {
      await result.current.run();
      await result.current.run();
    });
    expect(result.current.data).toEqual([
      { id: 10001, name: "Page:1" },
      { id: 10002, name: "Page:1" },
      { id: 20002, name: "Page:2" },
      { id: 20003, name: "Page:2" },
      { id: 10003, name: "Page:3" },
      { id: 10004, name: "Page:3" },
    ]);
  });
});
