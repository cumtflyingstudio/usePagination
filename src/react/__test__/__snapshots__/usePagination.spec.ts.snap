// Vitest Snapshot v1

exports[`usePagination > init should be with refresh once 1`] = `
[
  {
    "id": 1000,
    "item": {
      "id": 1000,
      "name": "Page:1",
    },
  },
  {
    "id": 1001,
    "item": {
      "id": 1001,
      "name": "Page:1",
    },
  },
]
`;

exports[`usePagination > test run once 1`] = `
[
  {
    "id": 1000,
    "item": {
      "id": 1000,
      "name": "Page:1",
    },
  },
  {
    "id": 1001,
    "item": {
      "id": 1001,
      "name": "Page:1",
    },
  },
  {
    "id": 1000,
    "item": {
      "id": 1000,
      "name": "Page:2",
    },
  },
  {
    "id": 1001,
    "item": {
      "id": 1001,
      "name": "Page:2",
    },
  },
]
`;
